client/keystore/src/lib.rs

```diff
@@ -276,11 +277,87 @@ impl Store {
 		buf.push(key_type + key.as_str());
 		Some(buf)
 	}
+
+	/// Signs a message with the private key that matches
+	/// the public key passed.
+	fn _sign_with(
+		&self,
+		id: KeyTypeId,
+		key: &CryptoTypePublicPair,
+		msg: &[u8],
+	) -> std::result::Result<Vec<u8>, String> {
+		match key.0 {
+			ed25519::ED25519_CRYPTO_ID => {
+				let pub_key = ed25519::Public::from_slice(key.1.as_slice());
+				let key_pair: ed25519::Pair = self
+					.key_pair_by_type::<ed25519::Pair>(&pub_key, id)
+					.map(Into::into)
+					.map_err(|e| e.to_string())?;
+				return Ok(<[u8; 64]>::from(key_pair.sign(msg)).to_vec());
+			}
+			sr25519::SR25519_CRYPTO_ID => {
+				let pub_key = sr25519::Public::from_slice(key.1.as_slice());
+				let key_pair: sr25519::Pair = self
+					.key_pair_by_type::<sr25519::Pair>(&pub_key, id)
+					.map(Into::into)
+					.map_err(|e| e.to_string())?;
+				return Ok(<[u8; 64]>::from(key_pair.sign(msg)).to_vec());
+			}
+			_ => Err(String::from("Key kind invalid"))
+		}
+	}
 }
 
 impl BareCryptoStore for Store {
+	fn get_supported_keys(&self, id: KeyTypeId, keys: Vec<CryptoTypePublicPair>) -> std::result::Result<Vec<CryptoTypePublicPair>, String> {
+		let ed25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<ed25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+		let sr25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<sr25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+
+		Ok(keys.iter().filter_map(|k| {
+			match k.0 {
+				sr25519::SR25519_CRYPTO_ID if sr25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
+				ed25519::ED25519_CRYPTO_ID if ed25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
+				_ => None
+			}
+		}).cloned().collect::<Vec<_>>())
```

Comment:
do we really have to clone here? How come?

```suggestion
		}).collect::<Vec<_>>())
```

Reply:
Without `cloned()`:
```
error[E0308]: mismatched types
   --> client/keystore/src/lib.rs:320:6
    |
320 |           Ok(keys.iter().filter_map(|k| {
    |  ____________^
321 | |             match k.0 {
322 | |                 sr25519::SR25519_CRYPTO_ID if sr25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
323 | |                 ed25519::ED25519_CRYPTO_ID if ed25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
324 | |                 _ => None
325 | |             }
326 | |         }).collect::<Vec<_>>())
    | |______________________________^ expected tuple, found reference
    |
    = note: expected struct `std::vec::Vec<(sp_core::crypto::CryptoTypeId, std::vec::Vec<u8>)>`
               found struct `std::vec::Vec<&(sp_core::crypto::CryptoTypeId, std::vec::Vec<u8>)>`
```